File: Enigma/EnigmaMachine.cs
using System;
using System.Collections.Generic;

namespace Enigma;

// Disables warning for reference of a potential null values
#pragma warning disable CS8602

public class EnigmaMachine
{
    private const string alphabet = "ABCDEFGHIJKLMNOPQRSTUVWXYZ";
    public readonly string[] rotorSets = { "EKMFLGDQVZNTOWYHXUSPAIBRCJ", "AJDKSIRUXBLHWTMCQGZNPYFVOE", "BDFHJLCPRTXVZNYEIWGAKMUSQO", "JFPGQCNLSVAUEWKTORDMBZIXYH", "FRAOEGYVICSBWZKQPJXTHLUDMN" };
    private List<string> rotors = new List<string>();
    private string? reflectorPlate;
    private Dictionary<char, char>? plugboardConnections;
    private List<int>? rotorPositions;

    public void ChangeMachineSettings(List<int> chosenRotors, string reflectorSet, List<Tuple<char, char>> plugboardSettings, List<int> rotorStartPositions)
    {
        foreach (int index in chosenRotors)
        {
            rotors.Add(rotorSets[index]);
        }
        reflectorPlate = reflectorSet;
        plugboardConnections = CreatePlugboard(plugboardSettings);
        rotorPositions = rotorStartPositions;
    }

    private Dictionary<char, char> CreatePlugboard(List<Tuple<char, char>> plugboardSettings)
    {
        Dictionary<char, char> plugboard = new Dictionary<char, char>();
        foreach (char letter in alphabet)
        {
            plugboard.Add(letter, letter);
        }

        for (int i = 0; i < plugboardSettings.Count; i++)
        {
            char a = plugboardSettings[i].Item1;
            char b = plugboardSettings[i].Item2;
            plugboard[a] = b;
            plugboard[b] = a;
        }

        return plugboard;
    }

    private void RotateRotors()
    {
        rotorPositions[0] = (rotorPositions[0] + 1) % 26;
        if (rotorPositions[0] != 0) return;

        rotorPositions[1] = (rotorPositions[1] + 1) % 26;
        if (rotorPositions[1] != 0) return;

        rotorPositions[2] = (rotorPositions[2] + 1) % 26;
    }

    private char EncryptLetter(char letter)
    {
        letter = plugboardConnections[letter];

        for (int i = 0; i < rotors.Count; i++)
        {
            string currentRotor = rotors[i];
            int index = ((alphabet.IndexOf(letter)) + rotorPositions[i]) % 26;
            letter = Convert.ToChar(currentRotor[index]);
        }

        letter = reflectorPlate[alphabet.IndexOf(letter)];

        for (int i = rotors.Count - 1; i >= 0; i--)
        {
            string currentRotor = rotors[i];
            int index = ((currentRotor.IndexOf(letter) - rotorPositions[i]) + 26) % 26;
            letter = alphabet[index];
        }

        letter = plugboardConnections[letter];

        RotateRotors();

        return letter;
    }

    public string EncryptMessage(string message)
    {
        List<char> encryptedCharacters = new List<char>();
        string encryptedMessage = "";

        foreach (char letter in message)
        {
            if (alphabet.Contains(letter))
            {
                char encryptedLetter = EncryptLetter(letter);
                encryptedCharacters.Add(encryptedLetter);
            }
            else
            {
                encryptedCharacters.Add(letter);
            }
        }

        foreach (char character in encryptedCharacters)
        {
            encryptedMessage += character;
        }

        return encryptedMessage;
    }
}

File: Java/src/com/cryptography/logic/enigma/EnigmaMachine.java
package com.cryptography.logic.enigma;

import com.cryptography.logic.Constants;
import com.cryptography.logic.Tuple2;

import java.util.ArrayList;
import java.util.HashMap;

public abstract class EnigmaMachine {
    protected ArrayList<String> _rotors = new ArrayList<>();
    protected String _reflectorPlate;
    protected HashMap<Character, Character> _plugboardConnections;
    protected Integer[] _rotorPositions;
    protected Integer[] _chosenRotors;

    protected void changeMachineSettings(String reflectorSet, ArrayList<Tuple2<Character, Character>> plugboardSettings, Integer[] rotorStartPositions, Integer[] chosenRotors) {
        _reflectorPlate = reflectorSet;
        _plugboardConnections = createPlugboard(plugboardSettings);
        _rotorPositions = rotorStartPositions;
        _chosenRotors = chosenRotors;
    }

    protected HashMap<Character, Character> createPlugboard(ArrayList<Tuple2<Character, Character>> plugboardSettings) {
        HashMap<Character, Character> plugboard = new HashMap<>();

        for (char letter : Constants.AlphabetList) {
            plugboard.put(letter, letter);
        }

        for (int i = 0; i < plugboardSettings.size(); i++) {
            char a = plugboardSettings.get(i).Item1();
            char b = plugboardSettings.get(i).Item2();
            plugboard.replace(a, b);
            plugboard.replace(b, a);
        }

        return plugboard;
    }

    protected void rotateRotors() {
        _rotorPositions[0] = (_rotorPositions[0] + 1) % 26;
        if (!isNotched(0)) return;

        _rotorPositions[1] = (_rotorPositions[1] + 1) % 26;
        if (!isNotched(1)) return;

        _rotorPositions[2] = (_rotorPositions[2] + 1) % 26;
    }

    public String encryptMessage(String message) {
        ArrayList<Character> encryptedCharacters = new ArrayList<>();
        StringBuilder encryptedMessage = new StringBuilder();

        for (char letter : message.toCharArray()) {
            if (Constants.AlphabetList.contains(letter)) {
                char encryptedLetter = encryptLetter(letter);
                encryptedMessage.append(encryptedLetter);
            }
            else
                encryptedMessage.append(letter);
        }

        return encryptedMessage.toString();
    }

    protected abstract char encryptLetter(char letter);

    protected boolean isNotched(int rotorIndex) {
        int rotor = _chosenRotors[rotorIndex];
        int position = _rotorPositions[rotorIndex];

        return switch (rotor) {
            case 0 -> position == 17;
            case 1 -> position == 5;
            case 2 -> position == 22;
            case 3 -> position == 10;
            case 4 -> position == 0;
            case 5, 6, 7 -> position == 0 || position == 13;
            default -> false;
        };
    }
}
